### PR TITLE
fix(sdk): fix resolving of default value for floating point parameter. Fixes #4222

### LINKED ISSUE #4222: Invalid format of float in scientific notation in workflow specification
Reopen #3806

### Environment:
KFP version:
Build commit: 1168b08

KFP SDK version:
kfp 1.0.0rc5
kfp-server-api 1.0.0rc5

/kind bug
/area sdk

### PR DESCRIPTION
**Pull request by TheFreaky**
**Description of your changes:**
Fix #4222

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [x] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

### CODE REVIEW

**Comment by TheFreaky** [comment]
Description of issue and fix:
```python
>>> from kfp import dsl, compiler
... @dsl.pipeline(name="test", description="test")
... def train_pipeline(
...     schedule_1_learning_rate: float = 0.0002,
...     schedule_2_learning_rate: float = 0.00002,
...     schedule_3_learning_rate: float = 2.e-6,
... ):
...     pass
... 
>>> import inspect
>>> inspect.getfullargspec(train_pipeline)
FullArgSpec(args=['schedule_1_learning_rate', 'schedule_2_learning_rate', 'schedule_3_learning_rate'], varargs=None, varkw=None, defaults=(0.0002, 2e-05, 2e-06), kwonlyargs=[], kwonlydefaults=None, annotations={'schedule_1_learning_rate': <class 'float'>, 'schedule_2_learning_rate': <class 'float'>, 'schedule_3_learning_rate': <class 'float'>})
>>> inspect.getfullargspec(train_pipeline).defaults
(0.0002, 2e-05, 2e-06)
>>> import yaml
>>> yaml.dump({'schedule_2_learning_rate': '2e-05'})
'schedule_2_learning_rate: 2e-05\n'
>>> yaml.dump({'schedule_2_learning_rate': '2.e-05'})
"schedule_2_learning_rate: '2.e-05'\n"
>>>
>>> # inspect.getfullargspec() retrieve float in form `2e-05`, but PyYAML requires `2.e-05`.
>>>
>>> import re
>>> yaml.Dumper.add_implicit_resolver(
...             u'tag:yaml.org,2002:float',
...             re.compile(u'''^(?:
...              [-+]?(?:[0-9][0-9_]*)\\.[0-9_]*(?:[eE][-+]?[0-9]+)?
...             |[-+]?(?:[0-9][0-9_]*)(?:[eE][-+]?[0-9]+)
...             |\\.[0-9_]+(?:[eE][-+][0-9]+)?
...             |[-+]?[0-9][0-9_]*(?::[0-5]?[0-9])+\\.[0-9_]*
...             |[-+]?\\.(?:inf|Inf|INF)
...             |\\.(?:nan|NaN|NAN))$''', re.X),
...             list(u'-+0123456789.'))
>>> yaml.dump({'schedule_2_learning_rate': '2e-05'})
"schedule_2_learning_rate: '2e-05'\n"
```

**Comment by TheFreaky** [comment]
@googlebot I signed it!

**Comment by Ark-kun** [comment]
Thank you for reopening the issue and contributing the fix.

I think the root issue is that Python's yaml module and Go's yaml module have different opinions about how the floating point numbers can look. So when python writes something, Go may be unable to read it.

But what's triggering this issue? I think there are different opinions here:

It seems to me that your fix assumes that the issue is in how python's yaml package serializes floats (or how Go reads floats).

I think that, instead, the issue is in how python's yaml package serializes strings (or how Go reads floats).

It seems to me that your `getfullargspec` example might not be demonstrating the issue. The raw default values are no longer just added to the arguments map. They're serialized first. But it looks like this does not fully fix the issue.

I've checked that my old fix has properly fixed the default values serialization to strings:

```python
import kfp
from kfp import components
chicago_taxi_dataset_op = components.load_component_from_url("https://raw.githubusercontent.com/kubeflow/pipelines/e3337b8bdcd63636934954e592d4b32c95b49129/components/datasets/Chicago%20Taxi/component.yaml")
def train_pipeline(
    schedule_1_learning_rate: float = 0.0002,
    schedule_2_learning_rate: float = 0.00002,
    schedule_3_learning_rate: float = 0.000002,
):
    chicago_taxi_dataset_op()

print(kfp.compiler.Compiler()._compile(train_pipeline)["spec"]["arguments"])
# {'parameters': [{'name': 'schedule_1_learning_rate', 'value': '0.0002'}, {'name': 'schedule_2_learning_rate', 'value': '2e-05'}, {'name': 'schedule_3_learning_rate', 'value': '2e-06'}]}
```
As you see, all arguments are strings, not floats.

Unfortunately Python and Go have different opinions about those strings. Go considers them to be float representations while Python does not. So python does not use quoting.

Let's check the following example:

```python
import yaml
print(yaml.dump({"a": "0.0002", "b": "2e-05"}))
# a: '0.0002'
# b: 2e-05
```

As you see, `0.0002` is quoted, but `2e-05` isn't. There was the same issue with booleans. The fix for that can be seen in the dump_yaml method.

I think your code fixes the yaml dumping by changing yaml parsing to recognize those strings as possible float representations and quoting them to prevent confusion. You might want to add a small comment there describing how this implicit *deserializer* fixed *serialization*. Otherwise it might be a bit confusing.

I guess the only remaining question is about the completeness of the regexp - how well does it correspond to other yaml parsers.
